app/com/linkedin/drelephant/spark/legacydata/SparkJobProgressData.java

```diff
@@ -0,0 +1,273 @@
+    public int numFailedTasks = 0;
+
+    /* Stages */
+    public int numActiveStages = 0;
+    // This needs to be a set instead of a simple count to prevent double-counting of rerun stages:
+    public final Set<Integer> completedStageIndices = new HashSet<Integer>();
+    public int numSkippedStages = 0;
+    public int numFailedStages = 0;
+
+    public void addStageId(int stageId) {
+      stageIds.add(stageId);
+    }
+
+    public double getFailureRate() {
+      return SparkJobProgressData.getFailureRate(numCompletedTasks, numFailedTasks);
+    }
+
+    public String toString() {
+      return String.format("{jobId:%s, jobGroup:%s, startTime:%s, endTime:%s, numTask:%s, numActiveTasks:%s, "
+              + "numCompletedTasks:%s, numSkippedTasks:%s, numFailedTasks:%s, numActiveStages:%s, "
+              + "completedStageIndices:%s, stages:%s, numSkippedStages:%s, numFailedStages:%s}", jobId, jobGroup,
+          startTime, endTime, numTasks, numActiveTasks, numCompletedTasks, numSkippedTasks, numFailedTasks,
+          numActiveStages, getListString(completedStageIndices), getListString(stageIds), numSkippedStages,
+          numFailedStages);
+    }
+  }
+
+  public static class StageInfo {
+    public int numActiveTasks;
+    public int numCompleteTasks;
+    public final Set<Integer> completedIndices = new HashSet<Integer>();
+    public int numFailedTasks;
+
+    // Total accumulated executor runtime
+    public long executorRunTime;
+    // Total stage duration
+    public long duration;
+
+    // Note, currently calculating I/O speed on stage level does not make sense
+    // since we do not have information about specific I/O time.
+    public long inputBytes = 0;
+    public long outputBytes = 0;
+    public long shuffleReadBytes = 0;
+    public long shuffleWriteBytes = 0;
+    public long memoryBytesSpilled = 0;
+    public long diskBytesSpilled = 0;
+
+    public String name;
+    public String description;
+
+    public double getFailureRate() {
+      return SparkJobProgressData.getFailureRate(numCompleteTasks, numFailedTasks);
+    }
+
+    // TODO: accumulables info seem to be unnecessary, might might be useful later on
+    // sample code from Spark source: var accumulables = new HashMap[Long, AccumulableInfo]
+
+    @Override
+    public String toString() {
+      return String.format("{numActiveTasks:%s, numCompleteTasks:%s, completedIndices:%s, numFailedTasks:%s,"
+              + " executorRunTime:%s, inputBytes:%s, outputBytes:%s, shuffleReadBytes:%s, shuffleWriteBytes:%s,"
+              + " memoryBytesSpilled:%s, diskBytesSpilled:%s, name:%s, description:%s}",
+          numActiveTasks, numCompleteTasks, getListString(completedIndices), numFailedTasks, executorRunTime,
+          inputBytes, outputBytes, shuffleReadBytes, shuffleWriteBytes, memoryBytesSpilled, diskBytesSpilled, name,
+          description);
+    }
+  }
+
+  private static String getListString(Collection collection) {
```

Comment:
fixed to use Collection<Integer>. I dont see collection.toString producing the same formatted output and hence keeping that part as is.